File: sdk/typescript/tests/e2e/coding-pipeline.test.ts
/**
 * TS 线一体化 coding harness 端到端测试
 *
 * 与 Go 线 test/e2e/coding_pipeline_test.go 镜像：
 * 计划（LLMPlanner 分解 DAG）→ 编写（filesystem 写文件）→ 测试（shell 校验）
 * → 审查（LLMReflector 批评）→ 发布（shell 驱动 git add/commit/tag），
 * 全部由脚本化 Provider 驱动，不依赖真实 LLM 与网络。
 */
import { describe, it, expect } from 'vitest';
import * as fs from 'node:fs';
import * as os from 'node:os';
import * as path from 'node:path';
import { execFileSync } from 'node:child_process';
import { ReActAgent } from '../../src/agent/react-loop.js';
import { MockProvider } from '../../src/llm/provider.js';
import { ToolRegistry } from '../../src/tools/registry.js';
import { FileSystemTool } from '../../src/tools/builtin/filesystem.js';
import { ShellTool } from '../../src/tools/builtin/shell.js';
import { LLMPlanner } from '../../src/agent/planning.js';
import { LLMReflector } from '../../src/agent/reflection.js';
import type { ToolCall, CompletionRequest, ToolCallRequest, CompletionResponse, ToolCallResponse } from '../../src/types.js';

// ===== 脚本化 Provider：Complete 与 CallTools 双队列，时序确定 =====

class ScriptedProvider extends MockProvider {
  completeQueue: string[] = [];
  toolCallQueue: { content: string; toolCalls: ToolCall[] }[] = [];

  withResponse(content: string): this {
    this.completeQueue.push(content);
    return this;
  }

  withToolResponse(toolCalls: ToolCall[], content = ''): this {
    this.toolCallQueue.push({ content, toolCalls });
    return this;
  }

  override async complete(_req: CompletionRequest): Promise<CompletionResponse> {
    const content = this.completeQueue.shift() ?? 'default complete';
    return {
      id: 'scripted', content, role: 'assistant',
      usage: { promptTokens: 1, completionTokens: 1, totalTokens: 2 },
    };
  }

  override async callTools(_req: ToolCallRequest): Promise<ToolCallResponse> {
    const next = this.toolCallQueue.shift() ?? { content: 'fallback', toolCalls: [] };
    return {
      content: next.content, toolCalls: next.toolCalls,
      usage: { promptTokens: 1, completionTokens: 1, totalTokens: 2 },
    };
  }
}

// ===== 协议常量 =====

// 根计划：编写 → 测试 → 审查 → 发布 依赖链
const pipelinePlan = JSON.stringify([
  { id: '1', description: '编写：创建 hello.ts 文件', depends_on: [] },
  { id: '2', description: '测试：检查工作区确认文件已生成', depends_on: ['1'] },
  { id: '3', description: '审查：评估代码质量并给出结论', depends_on: ['2'] },
  { id: '4', description: '发布：git 提交并打标签 v1.0.0', depends_on: ['3'] },
]);

// 批评应答：severity low 低于阈值 high，不触发改写
const lowCritique = JSON.stringify({ issues: [], severity: 'low', corrections: [] });

// ===== 辅助 =====

function git(workdir: string, ...args: string[]): string {
  return execFileSync('git', args, { cwd: workdir, encoding: 'utf-8' });
}

function hasGit(): boolean {
  try {
    execFileSync('git', ['--version'], { encoding: 'utf-8' });
    return true;
  } catch {
    return false;
  }
}

describe('TS coding pipeline 端到端', () => {
  it('计划→编写→测试→审查→发布全流程打通', async () => {
    if (!hasGit()) return; // 无 git 环境跳过

    const workdir = fs.mkdtempSync(path.join(os.tmpdir(), 'ts-coding-pipeline-'));
    try {
      git(workdir, 'init', '-b', 'main');
      git(workdir, 'config', 'user.email', 'e2e@example.com');
      git(workdir, 'config', 'user.name', 'e2e-bot');

      const provider = new ScriptedProvider();
      provider.withResponse(pipelinePlan); // 根任务 Planning：分解为 4 个子任务

      const fileContent = 'export const hello = (): string => "Hello from AgentPrimordia TS harness!";\n';

      // 子任务1：编写（filesystem 写文件）
      provider
        .withToolResponse([{
          id: 'call_write', name: 'filesystem',
          arguments: JSON.stringify({ action: 'write', path: 'hello.ts', content: fileContent }),
        }])
        .withToolResponse([], '已创建 hello.ts')
        .withResponse(lowCritique);

      // 子任务2：测试（shell 校验工作区）
      provider
        .withToolResponse([{
          id: 'call_check', name: 'shell',
          arguments: JSON.stringify({ command: 'git status --short', cwd: workdir }),
        }])
        .withToolResponse([], '工作区检查通过')
        .withResponse(lowCritique);

      // 子任务3：审查（无工具，直接结论）
      provider
        .withToolResponse([], '审查通过：代码结构清晰')
        .withResponse(lowCritique);

      // 子任务4：发布（git add + commit 同轮双调用，再 tag）
      provider
        .withToolResponse([
          { id: 'call_add', name: 'shell', arguments: JSON.stringify({ command: 'git add .', cwd: workdir }) },
          {
            id: 'call_commit', name: 'shell',
            arguments: JSON.stringify({ command: 'git commit -m "feat: add hello.ts"', cwd: workdir }),
          },
        ])
        .withToolResponse([{
          id: 'call_tag', name: 'shell',
          arguments: JSON.stringify({ command: 'git tag v1.0.0', cwd: workdir }),
        }])
        .withToolResponse([], '发布完成：v1.0.0')
        .withResponse(lowCritique);

      // ===== 装配一体化 harness =====
      const registry = new ToolRegistry();
      registry.register(new FileSystemTool({ rootDir: workdir }));
      registry.register(new ShellTool({ workingDir: workdir }));

      const agent = new ReActAgent({
        name: 'coding-agent',
        model: provider,
        toolkit: registry,
        systemPrompt: '你是一个全自动编码 Agent，按计划完成编写、测试、审查与发布。',
        maxTurns: 8,
        planner: new LLMPlanner(provider),
        reflector: new LLMReflector(provider),
        reflectionSeverityThreshold: 'high',
      });

      const resp = await agent.run('创建 hello.ts，验证工作区，审查代码，然后提交并打标签 v1.0.0');

      // ===== 断言 =====
      expect(resp.content).toBe('发布完成：v1.0.0');
      expect(fs.readFileSync(path.join(workdir, 'hello.ts'), 'utf-8')).toBe(fileContent);
      expect(git(workdir, 'log', '--oneline')).toContain('feat: add hello.ts');
      expect(git(workdir, 'tag', '-l')).toContain('v1.0.0');
      // write + status + add + commit + tag = 5 次工具调用
      expect(resp.metrics.totalTools).toBe(5);
      // 批评队列被 4 个子任务各消费一次
      expect(provider.completeQueue.length).toBe(0);
    } finally {
      fs.rmSync(workdir, { recursive: true, force: true });
    }
  }, 30_000);
});

File: agentprimordia/test/e2e/coding_pipeline_test.go
// Package e2e 提供 AgentPrimordia 端到端集成测试。
//
// coding_pipeline_test.go 验证「一体化 coding harness」全流程：
// 计划（Planner 分解 DAG）→ 编写（filesystem 写文件）→ 测试（shell 校验）
// → 审查（Reflector 批评最终输出）→ 发布（git add/commit/tag），
// 全部由 MockLLM 脚本化驱动，不依赖真实 LLM 与网络。
package e2e

import (
	"context"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"testing"

	gitplugin "agentprimordia/ecosystem/plugins/git"
	"agentprimordia/internal/llm"
	ap "agentprimordia/pkg"
)

// singleSubtaskPlan 是子任务 runLoop 再次触发 Planning 时的应答：
// 仅含 1 个子任务时引擎不再走 DAG 分支（需 >1 才 executePlan），
// 避免无限递归分解。
const singleSubtaskPlan = `[{"id": "1", "description": "直接执行当前步骤", "depends_on": []}]`

// lowCritique 是 Reflector.Critique 的应答：严重度 low 低于阈值 high，
// 不触发 Improve 二次改写，保证最终输出可断言。
const lowCritique = `{"issues": [], "severity": "low", "corrections": []}`

// pipelinePlan 是根任务的分解计划：编写 → 测试 → 审查 → 发布 依赖链
const pipelinePlan = `[
  {"id": "1", "description": "编写：创建 hello.go 文件", "depends_on": []},
  {"id": "2", "description": "测试：检查工作区确认文件已生成", "depends_on": ["1"]},
  {"id": "3", "description": "审查：评估代码质量并给出结论", "depends_on": ["2"]},
  {"id": "4", "description": "发布：git 提交并打标签 v1.0.0", "depends_on": ["3"]}
]`

// initGitRepo 在指定目录初始化带用户配置的 Git 仓库
func initGitRepo(t *testing.T, dir string) {
	t.Helper()
	for _, args := range [][]string{
		{"init", "-b", "main"},
		{"config", "user.email", "e2e@example.com"},
		{"config", "user.name", "e2e-bot"},
	} {
		cmd := exec.Command("git", args...)
		cmd.Dir = dir
		if out, err := cmd.CombinedOutput(); err != nil {
			t.Fatalf("git %v failed: %v\n%s", args, err, out)
		}
	}
}

// gitOutput 在目录中执行 git 命令并返回输出
func gitOutput(t *testing.T, dir string, args ...string) string {
	t.Helper()
	cmd := exec.Command("git", args...)
	cmd.Dir = dir
	out, err := cmd.CombinedOutput()
	if err != nil {
		t.Fatalf("git %v failed: %v\n%s", args, err, out)
	}
	return string(out)
}

// mustJSONString 将字符串编码为 JSON 字符串字面量（含引号），
// 用于在 tool 参数中内嵌文件内容/路径
func mustJSONString(s string) string {
	var sb strings.Builder
	sb.WriteByte('"')
	for _, r := range s {
		switch r {
		case '"':
			sb.WriteString(`\"`)
		case '\\':
			sb.WriteString(`\\`)
		case '\n':
			sb.WriteString(`\n`)
		case '\t':
			sb.WriteString(`\t`)
		default:
			sb.WriteRune(r)
		}
	}
	sb.WriteByte('"')
	return sb.String()
}

// scriptSubtask 按子任务时序追加 Mock 响应：
// Complete 队列 = 计划检查（单子任务降级）→ 结论文本 → Critique；
// CallTools 队列 = 若干 tool 调用轮 + 末尾一个空响应（触发回退 Complete 给出结论）。
func scriptSubtask(mock *llm.MockLLM, conclusion string, toolRounds [][]llm.FunctionCall) {
	mock.WithResponse(singleSubtaskPlan)
	for _, round := range toolRounds {
		mock.WithToolResponse(round)
	}
	mock.WithToolResponse(nil) // 空 tool 响应 → syncReasoning 回退 Complete
	mock.WithResponse(conclusion)
	mock.WithResponse(lowCritique)
}

// TestCodingPipeline_EndToEnd 验证计划→编写→测试→审查→发布全流程打通。
//
// 执行时序（每个子任务都会重新进入 runLoop，startTurn=0 再次触发 Planning，
// 用单子任务计划使其降级为普通 ReAct 循环）：
//
//	根 runLoop: Planner 分解为 4 子任务 → executePlan DAG 分层执行
//	子任务1 编写: filesystem.write
//	子任务2 测试: shell(git status)
//	子任务3 审查: 无 tool 直接给结论
//	子任务4 发布: git add + commit（同一轮两个调用）→ git tag
//	每个子任务完成路径: Reflector.Critique（low → 不改写）
func TestCodingPipeline_EndToEnd(t *testing.T) {
	if _, err := exec.LookPath("git"); err != nil {
		t.Skip("git not available")
	}

	workdir := t.TempDir()
	initGitRepo(t, workdir)

	mock := llm.NewMockLLM(t)
	mock.WithResponse(pipelinePlan) // 根任务 Planning：分解为 4 个子任务

	fileContent := "package main\n\nfunc main() {}\n"

	// 子任务1：编写（filesystem 写文件）
	scriptSubtask(mock, "已创建 hello.go", [][]llm.FunctionCall{{
		{
			ID:        "call_write",
			Name:      "filesystem",
			Arguments: `{"action": "write", "path": "hello.go", "content": ` + mustJSONString(fileContent) + `}`,
		},
	}})

	// 子任务2：测试（shell 检查工作区）
	scriptSubtask(mock, "工作区检查通过", [][]llm.FunctionCall{{
		{
			ID:        "call_check",
			Name:      "shell",
			Arguments: `{"action": "execute", "command": "git status --short", "workdir": ` + mustJSONString(workdir) + `}`,
		},
	}})

	// 子任务3：审查（无 tool，直接结论）
	scriptSubtask(mock, "审查通过：代码结构清晰", nil)

	// 子任务4：发布（git add+commit 同轮双调用，再 tag）
	scriptSubtask(mock, "发布完成：v1.0.0", [][]llm.FunctionCall{
		{
			{
				ID:        "call_add",
				Name:      "git_tool",
				Arguments: `{"action": "add", "args": ["."], "workdir": ` + mustJSONString(workdir) + `}`,
			},
			{
				ID:        "call_commit",
				Name:      "git_tool",
				Arguments: `{"action": "commit", "message": "feat: add hello.go", "workdir": ` + mustJSONString(workdir) + `}`,
			},
		},
		{
			{
				ID:        "call_tag",
				Name:      "git_tool",
				Arguments: `{"action": "tag", "name": "v1.0.0", "message": "release v1.0.0", "workdir": ` + mustJSONString(workdir) + `}`,
			},
		},
	})

	// ===== 装配 Agent：Planner + Reflector + filesystem/shell/git 工具 =====
	registry := ap.NewToolRegistry()
	fs, err := ap.NewFileSystem(workdir)
	if err != nil {
		t.Fatalf("创建 filesystem 工具失败: %v", err)
	}
	if err := registry.Register(fs); err != nil {
		t.Fatalf("注册 filesystem 失败: %v", err)
	}
	if err := registry.Register(ap.NewShell()); err != nil {
		t.Fatalf("注册 shell 失败: %v", err)
	}
	if err := registry.RegisterPlugin(gitplugin.New()); err != nil {
		t.Fatalf("注册 git 插件失败: %v", err)
	}

	codingAgent, err := ap.NewAgent("coding-agent",
		"你是一个全自动编码 Agent，按计划完成编写、测试、审查与发布。",
		mock,
		ap.WithMaxTurns(8),
		ap.WithToolkit(registry),
		ap.WithCognition(ap.CognitionConfig{
			Planner:                     ap.NewLLMPlanner(mock),
			Reflector:                   ap.NewLLMReflector(mock),
			ReflectionSeverityThreshold: "high",
		}),
	)
	if err != nil {
		t.Fatalf("创建 Agent 失败: %v", err)
	}

	resp, err := codingAgent.Run(context.Background(),
		ap.UserMessage("创建 hello.go，验证工作区，审查代码，然后提交并打标签 v1.0.0"))
	if err != nil {
		t.Fatalf("Run 失败: %v", err)
	}

	// 最终输出 = 最后一个子任务（发布）的结论
	if resp.Content != "发布完成：v1.0.0" {
		t.Errorf("最终输出 = %q, want %q", resp.Content, "发布完成：v1.0.0")
	}

	// 编写环节：文件已落盘且内容正确
	got, err := os.ReadFile(filepath.Join(workdir, "hello.go"))
	if err != nil {
		t.Fatalf("hello.go 未创建: %v", err)
	}
	if string(got) != fileContent {
		t.Errorf("hello.go 内容 = %q, want %q", string(got), fileContent)
	}

	// 发布环节：提交与标签进入仓库
	if log := gitOutput(t, workdir, "log", "--oneline"); !strings.Contains(log, "feat: add hello.go") {
		t.Errorf("git log 缺少提交, got: %s", log)
	}
	if tags := gitOutput(t, workdir, "tag", "--list"); !strings.Contains(tags, "v1.0.0") {
		t.Errorf("git tag 缺少 v1.0.0, got: %s", tags)
	}

	// 指标：共执行 5 次 tool 调用（write/check/add/commit/tag）
	if resp.Metrics.TotalTools != 5 {
		t.Errorf("TotalTools = %d, want 5", resp.Metrics.TotalTools)
	}
}
